networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/GossipHandler.java

```diff
@@ -73,20 +73,32 @@ public SafeFuture<ValidationResult> apply(final MessageApi message) {
     }
     LOG.trace("Received message for topic {}", topic);
 
-    PubsubMessage pubsubMessage = message.getOriginalMessage();
-    if (!(pubsubMessage instanceof PreparedPubsubMessage)) {
+    final PubsubMessage pubsubMessage = message.getOriginalMessage();
+    if (!(pubsubMessage instanceof PreparedPubsubMessage gossipPubsubMessage)) {
       throw new IllegalArgumentException(
           "Don't know this PubsubMessage implementation: " + pubsubMessage.getClass());
     }
-    PreparedPubsubMessage gossipPubsubMessage = (PreparedPubsubMessage) pubsubMessage;
     return handler.handleMessage(gossipPubsubMessage.getPreparedMessage());
   }
 
-  public void gossip(final Bytes bytes) {
+  /**
+   * This method is designed to return a future that completes successfully when
+   *
+   * <p>1. The message is successfully to at least one peer
+   *
+   * <p>2. An error occurred
+   */
+  public SafeFuture<Void> gossip(final Bytes bytes) {
     LOG.trace("Gossiping {}: {} bytes", topic, bytes.size());
-    SafeFuture.of(publisher.publish(Unpooled.wrappedBuffer(bytes.toArrayUnsafe()), topic))
-        .finish(
-            () -> LOG.trace("Successfully gossiped message on {}", topic),
-            err -> LOG.debug("Failed to gossip message on " + topic, err));
+    return SafeFuture.of(publisher.publish(Unpooled.wrappedBuffer(bytes.toArrayUnsafe()), topic))
+        .handle(
+            (result, error) -> {
+              if (error != null) {
+                LOG.debug("Failed to gossip message on {}", topic, error);
```

Comment:
maybe add some piece of data with supplier?

Reply:
I'll refactor to move error handling up in the stack so we can have a better error handling, and maybe unify it with what we are already doing with Attestations.